pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,728 @@
+	}
+
+	// Create a selectorItem if it doesn't exist.
+	sItem, exists := i.selectorItems[sKey]
+	if !exists {
+		sItem = i.createSelectorItem(gItem)
+	}
+	sItem.groupItemKeys.Insert(gKey)
+}
+
+func (i *GroupEntityIndex) DeleteGroup(groupType string, name string) {
+	gKey := getGroupItemKey(groupType, name)
+
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	gItem, exists := i.groupItems[gKey]
+	if !exists {
+		return
+	}
+
+	// Delete the group from its associated selectorItem and groupItems.
+	i.deleteGroupFromSelectorItem(gItem.selectorItemKey, gKey)
+	delete(i.groupItems, gKey)
+}
+
+func (i *GroupEntityIndex) notify(selector string) {
+	sItem := i.selectorItems[selector]
+	for group := range sItem.groupItemKeys {
+		i.eventChan <- group
+	}
+}
+
+func (i *GroupEntityIndex) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting GroupEntityIndex")
+	for {
+		select {
+		case <-stopCh:
+			klog.Info("Stopping GroupEntityIndex")
+			return
+		case group := <-i.eventChan:
+			parts := strings.Split(group, "/")
+			groupType, name := parts[0], parts[1]
+			for _, handler := range i.eventHandlers[groupType] {
+				handler(name)
+			}
+		}
+	}
+}
+
+func (i *GroupEntityIndex) AddEventHandler(groupType string, handler eventHandler) {
+	i.eventHandlers[groupType] = append(i.eventHandlers[groupType], handler)
+}
+
+func (i *GroupEntityIndex) match(entityType entityType, label labels.Set, namespace string, sel *types.GroupSelector) bool {
+	objSelector := sel.PodSelector
+	if entityType == externalEntityType {
+		objSelector = sel.ExternalEntitySelector
+	}
+	if sel.Namespace != "" {
+		if sel.Namespace != namespace {
+			// Pods or ExternalEntities must be matched within the same Namespace.
+			return false
+		}
+		if objSelector != nil && !objSelector.Matches(label) {
+			// podSelector or externalEntitySelector exists but doesn't match the ExternalEntity or Pod's labels.
+			return false
+		}
+		return true
+	}
+	if sel.NamespaceSelector != nil {
+		if !sel.NamespaceSelector.Empty() {
+			namespaceLabels, exists := i.namespaceLabels[namespace]
+			if !exists {
+				return false
+			}
+			if !sel.NamespaceSelector.Matches(namespaceLabels) {
+				// Pod's Namespace do not match namespaceSelector.
+				return false
+			}
+		}
+		if objSelector != nil && !objSelector.Matches(label) {
+			// ExternalEntity or Pod's Namespace matches namespaceSelector but
+			// labels do not match the podSelector or externalEntitySelector.
+			return false
+		}
+		return true
+	}
+	if objSelector != nil {
+		// Selector only has a PodSelector/ExternalEntitySelector and no sel.Namespace.
+		// Pods/ExternalEntities must be matched from all Namespaces.
+		if !objSelector.Matches(label) {
+			// pod/ee labels do not match PodSelector/ExternalEntitySelector.
+			return false
+		}
+		return true
+	}
+	// TODO: if none of the above selector is set, it means matching all entities or none?
+	// This is only possible for ClusterGroup. From previous code, it means matching none.
```

Comment:
I think @Dyanngg should confirm. If one wants to select all Pods in all Namespaces, I assume we need an empty `PodSelector` and an empty `NamespaceSelector` (as opposed to `nil` ones)?

Reply:
This is an undocumented case so I think we should keep it consistent with previous releases (matches nothing). As you can see such semantics is quite implicit. That's why I'm proposing to make `namespaceSelector` a required field for ClusterGroup and ACNP appliedTo for the next API version. This would eliminate such ambiguity.

Reply:
The current code keeps previous behavior

Reply:
yes, and I would agree with the current implementation.